### DISCUSSION TITLE
[BUG] - File I/O Side Effect of rdDetermineBonds.DetermineConnectivity with `useHueckel=True` causes crashes in read-only filesystems (e.g., RO Jupyter Notebook Environments)

### DISCUSSION THREAD

**Discussion by coltonbh** [rdkit/rdkit#8091] [Development]
When running:

```python
rdDetermineBonds.DetermineConnectivity(
            mol,
            charge=charge,
            useHueckel=use_hueckel,
            useVdw=use_vdw,
            covFactor=cov_factor,
        )
```

If using `useHueckel=True` rdkit writes two files to disk: `run.out` and `nul`. When trying to run this function in read-only file system environments (e.g., a Jupyter Notebook with a RO filesystem) this causes the kernel to crash.

**Is there a way to suppress this side effect behavior?** 

Thanks so much! :pray:

**Comment by bp-kelley** [comment]
The code looks to need to be altered.  There are two bugs here:

It looks like opening "nul" is from windows, this is the null file descriptor like /dev/null on linux.  The other is always writing run.out to the current working directory.

The issues are in EHTTools.cpp

**Comment by bp-kelley** [comment]
I think if you change your working directory to /tmp during determineBonds it should work.  I can check that.

**Comment by bp-kelley** [comment]
Here is a simple workaround:

```
import tempfile

curdir = os.path.abspath(os.curdir)
with tempfile.TemporaryDirectory() as d:
    os.chdir(d)
    try:
        rdDetermineBonds.DetermineConnectivity(m, useHueckel=True)
        print(Chem.MolToSmiles(m))
    finally:
        os.chdir(curdir)
```

**Reply by coltonbh** [reply]
Thanks for this! Helps in the short-term :)

**Comment by bp-kelley** [comment]
Assigned as a bug in #8092

**Comment by greglandrum** [comment]
A general comment here: I think that you'll find that you get good results from `determineConnectivity()` with the defaults (i.e. with `useHueckel=False`). If that's not the case, I'd be curious to hear about examples.

**Reply by coltonbh** [reply]
Have many examples where the default `DetermineConnectivity` method (connect-the-dots) does not work and either the Van der Waals or Hueckel method are required. As an example, here's a simple catalyst/counter ion structure that fails with the default method, fails with Van der Walls, but succeeds with heuckel :)

```
33
qcio_charge=0 qcio_multiplicity=1 -69.15289332
Na -1.95508118079999993 -3.21324132980000021 1.53896580520000015
S  -3.89275832929999988 -1.58076735009999991 0.81939584129999998
C  -2.53513138819999995 -0.81028284390000005 0.14661341570000000
N  -1.33856741479999997 -1.31800502939999986 0.39425722610000002
C  -0.17317247140000000 -0.64575149310000002 0.22050338019999999
C  0.98194754660000005 -1.38737766510000005 -0.08650232899999999
C  2.21318999979999997 -0.77360101630000000 -0.19267318219999999
C  3.41035027499999988 -1.60276762809999984 -0.55888297570000001
F  4.56829388590000018 -1.09545825659999996 -0.10746955640000000
F  3.34197085899999990 -2.85966006649999960 -0.08605857990000000
F  3.55812451099999993 -1.72499682409999999 -1.89677523530000003
C  2.36799830780000020 0.59692731509999997 -0.00698557750000000
C  1.23482963960000003 1.32839873690000010 0.32135066210000002
C  1.32497427749999996 2.80634126620000002 0.56737979360000002
F  2.53306333319999988 3.32841588660000021 0.31570938300000001
F  0.45143860209999997 3.50170164089999991 -0.19221278500000000
F  1.03236739710000003 3.13576200490000012 1.84270006450000001
C  -0.00964710230000000 0.73705563019999998 0.43977885329999999
N  -2.71190938059999986 0.23924612219999999 -0.67059791489999998
C  -3.96936400520000010 0.81934775159999984 -1.09790186079999996
C  -3.98223726870000005 2.30705172979999951 -0.74989071979999999
C  -4.16032520749999968 0.59340399909999986 -2.59658677090000012
H  -4.76024325699999995 0.30426335869999999 -0.54382042890000004
H  0.88703995240000000 -2.45325563769999988 -0.23604020949999999
H  -4.92619732030000002 2.75070111289999986 -1.05330292360000000
H  -3.86465190050000018 2.44738550950000011 0.32261271650000001
H  -3.17306701540000002 2.82641398190000004 -1.26076661359999997
H  -3.37561839440000000 1.09489603460000007 -3.16052778049999983
H  -4.13242278219999992 -0.47033105939999997 -2.82427761159999946
H  -5.12157628220000039 0.98949048230000003 -2.91142237439999985
H  3.33209875400000000 1.06892492499999991 -0.10052968910000000
H  -0.85300414390000001 1.33994217100000013 0.74387626100000004
H  -1.89056893450000008 0.59649613140000002 -1.14082251370000010
```

**Comment by coltonbh** [comment]
Thanks all for the quick look at this! 

Would be great for any function that has side effects like this to be able to turn them off with an argument, or perhaps better to have no side effects unless specifically asked for via an argument.

Again, much thanks! :pray:

**Reply by greglandrum** [reply]
> 
> Would be great for any function that has side effects like this to be able to turn them off with an argument, or perhaps better to have no side effects unless specifically asked for via an argument.

This sometimes isn't possible when working with libraries that are 25+ years old. Still, PRs against the yaehmop code base are always welcome if this is something you feel strongly about

**Reply by coltonbh** [reply]
Understood! 

I appreciate all the work the maintainers do to keep `rdkit` a resource for the community ❤️. It seems to have reliable implementations for a lot of the basic cheminformatics operations I need (connectivity, smiles generation, alignments, rmsd cals, etc.); are there other more modern tools you'd recommend as alternatives?

**Reply by greglandrum** [reply]
> Understood!
> 
> I appreciate all the work the maintainers do to keep `rdkit` a resource for the community ❤️. It seems to have reliable implementations for a lot of the basic cheminformatics operations I need (connectivity, smiles generation, alignments, rmsd cals, etc.); are there other more modern tools you'd recommend as alternatives?

Oh, I wasn't talking about the RDKit there, we actively maintain the RDKit and try to modernize it. The extended Hueckel code that is used in `DetermineBonds`, on the other hand, is from when I was in grad school and wasn't originally written to be used as a library.

**Reply by coltonbh** [reply]
Thanks for all the work to keep things running smoothly! Much appreciated 🙏
